flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MonitorSource.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.util.ratelimit.RateLimitedSourceReader;
+import org.apache.flink.api.connector.source.util.ratelimit.RateLimiter;
+import org.apache.flink.api.connector.source.util.ratelimit.RateLimiterStrategy;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.core.memory.DataInputDeserializer;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Monitors an Iceberg table for changes */
+public class MonitorSource extends SingleThreadedIteratorSource<TableChange> {
+  private static final Logger LOG = LoggerFactory.getLogger(MonitorSource.class);
+
+  private final TableLoader tableLoader;
+  private final RateLimiterStrategy rateLimiterStrategy;
+  private final long maxReadBack;
+
+  /**
+   * Creates a {@link org.apache.flink.api.connector.source.Source} which monitors an Iceberg table
+   * for changes.
+   *
+   * @param tableLoader used for accessing the table
+   * @param rateLimiterStrategy limits the frequency the table is checked
+   * @param maxReadBack sets the number of snapshots read before stopping change collection
+   */
+  public MonitorSource(
+      TableLoader tableLoader, RateLimiterStrategy rateLimiterStrategy, long maxReadBack) {
+    Preconditions.checkNotNull(tableLoader, "Table loader should no be null");
+    Preconditions.checkNotNull(rateLimiterStrategy, "Rate limiter strategy should no be null");
+    Preconditions.checkArgument(maxReadBack > 0, "Need to read at least 1 snapshot to work");
+
+    this.tableLoader = tableLoader;
+    this.rateLimiterStrategy = rateLimiterStrategy;
+    this.maxReadBack = maxReadBack;
+  }
+
+  @Override
+  public Boundedness getBoundedness() {
+    return Boundedness.CONTINUOUS_UNBOUNDED;
+  }
+
+  @Override
+  public TypeInformation<TableChange> getProducedType() {
+    return TypeInformation.of(TableChange.class);
+  }
+
+  @Override
+  Iterator<TableChange> createIterator() {
+    return new SchedulerEventIterator(tableLoader, null, maxReadBack);
+  }
+
+  @Override
+  SimpleVersionedSerializer<Iterator<TableChange>> getIteratorSerializer() {
+    return new SchedulerEventIteratorSerializer(tableLoader, maxReadBack);
+  }
+
+  @Override
+  public SourceReader<TableChange, GlobalSplit<TableChange>> createReader(
+      SourceReaderContext readerContext) throws Exception {
+    RateLimiter rateLimiter = rateLimiterStrategy.createRateLimiter(1);
+    return new RateLimitedSourceReader<>(super.createReader(readerContext), rateLimiter);
```

Comment:
why don't we just let the enumerator leverage the coordinator single thread pool to schedule the action. Using rate limiter reader seems unnecessary. reader will be idle if there is no split.

```
    enumeratorContext.callAsync(
        this::discoverSplits,
        this::processDiscoveredSplits,
        0L,
        scanContext.monitorInterval().toMillis());
```

Reply:
then we probably don't need the iterator reader. enumerator just emit a global split when appropriate.

Reply:
Here are the requirements:
- We need to emit the `TableChange` events for a given timestamp in a single event record - we want to prevent unnecessary Maintenance Tasks scheduling
- We need to have state for the source - we want to avoid emitting duplications, even on state restore

I have considered the following:
- The split is a `Snapshot`, and the enumerator emits snapshots, reader reads the snapshot and emits `TableChange` based on a given snapshot - this breaks our first requirement, as it will generate multiple small `TableChange` events on startup, or when multiple commits arrive between enumeration intervals
- The split is basically a `TableChange` calculated by the enumerator, and the reader just emits the value it has received - I think this also breaks the first requirement on state restore. We might emit an old `TableChange` and a new `TableChange` immediately. This could again double trigger our Maintenance Tasks. How sensitive is the Enumerator for long running `TableChange` calculations?

Do I miss something, or do you have another idea?

Thanks, Peter

Reply:
I was thinking if exposing `RateLimiterStrategy` is intuitive to users. `RateLimterStrategy` usually is used to avoid overwhelming downstream operators and cause severe backpressure. In this context, it seems to be used to simulate/implement periodical table check for new commits. `SplitEnumeratorContext.callAsync` was designed for periodical action on coordinator. I thought it is more intuitive to ask users to configure `monitorInterval(Duration interval)`

Reply:
wondering how would users construct the rate limiter strategy for the monitor source?

```
    /**
     * Creates a {@code RateLimiterStrategy} that is limiting the number of records per second.
     *
     * @param recordsPerSecond The number of records produced per second. The actual number of
     *     produced records is subject to rounding due to dividing the number of produced records
     *     among the parallel instances.
     */
    static RateLimiterStrategy perSecond(double recordsPerSecond) {
        return parallelism -> new GuavaRateLimiter(recordsPerSecond / parallelism);
    }

    /**
     * Creates a {@code RateLimiterStrategy} that is limiting the number of records per checkpoint.
     *
     * @param recordsPerCheckpoint The number of records produced per checkpoint. This value has to
     *     be greater or equal to parallelism. The actual number of produced records is subject to
     *     rounding due to dividing the number of produced records among the parallel instances.
     */
    static RateLimiterStrategy perCheckpoint(int recordsPerCheckpoint) {
        return parallelism -> {
            int recordsPerSubtask = recordsPerCheckpoint / parallelism;
            checkArgument(
                    recordsPerSubtask > 0,
                    "recordsPerCheckpoint has to be greater or equal to parallelism. "
                            + "Either decrease the parallelism or increase the number of "
                            + "recordsPerCheckpoint.");
            return new GatedRateLimiter(recordsPerSubtask);
        };
    }
```

Reply:
I have a better understanding of the `IteratorSourceEnumerator` and `GlobalSplit` now. Basically, we are assigning unbounded/never-end singleton split to the iterator reader. I guess this probably ties to my early question on what is a GlobalSplit. Now I can tie them together.

My main suggestion would be just avoid exposing `RateLimiterStrategy` to users in the constructor. It is probably more user friendly to change it to  `Duration monitorInterval`. Internally, using rate limiter is an implementation detail.

Reply:
I would like to avoid exposing the `MonitorSource` and other operators altogether. I propose to provide a `TableMaintenance.Builder` for the users, which chains together the operators and parametrizing them. I would like to keep the internal details internal, and I agree with you that we should provide `Duration` based configuration for the users on the `TableMaintenance.Builder`

I envision the following usage:
```
TableMaintenance.builder(env, tableLoader)
   .uidPrefix("UidForState")
   .rateLimit(Duration.ofMinutes(10))
   .add(ExpireSnapshots.builder().scheduleOnCommit(10))
   .add(RewriteDataFiles.builder().scheduleOnFileNumber(3))
   .add(RewriteManifestFiles.builder().scheduleOnCommit(5))
   .add(DeleteOrphanFiles.builder(env).scheduleOnTime(Duration.ofDays(1)))
   .build();
```

WDYT?

Reply:
>  would like to avoid exposing the MonitorSource and other operators altogether. I propose to provide a TableMaintenance.Builder for the users, which chains together the operators and parametrizing them. 

that would make sense. I didn't see the `MonitorSource` marked as internal. thought it will be exposed to users.

> .rateLimit(Duration.ofMinutes(10))

I would recommend `monitorInterval(Duration)`

Reply:
I see. right now, the connector lives in the Iceberg repo. we are trying to keep internal classes non-public. Some internal classes like this are made public due to sub packages etc. We have been trying to explicitly mark them as `@Internal`. Do you think it makes sense to continue that convention?

Reply:
Added `@Internal` for all of the classes